light_curves/light_curve_classifier.md

```diff
@@ -165,147 +172,202 @@ df_lc['label'] = df_lc.label.str.replace('Lopez-Navas 22', 'CLAGN')
-#remove instrumental noise and CR (ground based)
-
-#keep track of how many rows this removes
-start_len = len(df_lc)
-
-#setup to collect the outlier thresholds per band to later reject
-nsigmaonmean= {}
+def sigmaclip_lightcurves(df_lc, sigmaclip_value = 10.0, include_plot = False):
+    """
+    Sigmaclip to remove bad values from the light curves; optionally plots histograms of uncertainties
+        to help determine sigmaclip_value from the data. 
+    
+    Parameters
+    ----------
+    df_lc: Pandas dataframe with light curve info
+   
+    sigmaclip_value: float
+        what value of sigma should be used to make the cuts
+
+    include_plot: bool
+        have the function plot histograms of uncertainties for each band
+        
+    Returns
+    --------
+    df_lc: MultiIndexDFObject with all  light curves
+        
+    """
+    #keep track of how many rows this removes
+    start_len = len(df_lc)
 
-#what value of sigma are we going to consider is an outlier?
-sigmaclip_value = 10.0
+    #setup to collect the outlier thresholds per band to later reject
+    nsigmaonmean= {}
 
-#create the figure and axes
-fig, axs = plt.subplots(5, 3, figsize = (12, 12))
+    if include_plot:
+        #create the figure and axes
+        fig, axs = plt.subplots(5, 3, figsize = (12, 12))
 
-# unpack all the axes subplots
-axe = axs.ravel()
+        # unpack all the axes subplots
+        axe = axs.ravel()
 
-#for each band
-for count, (bandname, singleband) in enumerate(df_lc.groupby("band")):
+    #for each band
+    for count, (bandname, singleband) in enumerate(df_lc.groupby("band")):
     
-    #use scipy sigmaclip to iteratively determine sigma on the dataset
-    clippedarr, lower, upper = sigmaclip(singleband.err, low = sigmaclip_value, high = sigmaclip_value)
+        #use scipy sigmaclip to iteratively determine sigma on the dataset
+        clippedarr, lower, upper = sigmaclip(singleband.err, low = sigmaclip_value, high = sigmaclip_value)
     
-    #store this value for later
-    nsigmaonmean[bandname] = upper
+        #store this value for later
+        nsigmaonmean[bandname] = upper
     
-    #plot distributions and print stddev
-    singleband.err.plot(kind = 'hist', bins = 30, subplots =True, ax = axe[count],label = bandname+' '+str(upper), legend=True)
-```
+        if include_plot:        
+            #plot distributions and print stddev
+            singleband.err.plot(kind = 'hist', bins = 30, subplots =True, ax = axe[count],label = bandname+' '+str(upper), legend=True)
 
-```{code-cell} ipython3
-#remove data that are outside the sigmaclip_value
-for bandname, cut in nsigmaonmean.items():
-    querystring = f'band == {bandname!r} & err > {cut}'
-    df_lc = df_lc.drop(df_lc.query(querystring).index)
-```
+    #remove data that are outside the sigmaclip_value
+    for bandname, cut in nsigmaonmean.items():
+        querystring = f'band == {bandname!r} & err > {cut}'
+        df_lc = df_lc.drop(df_lc.query(querystring).index)
```

Comment:
You can avoid changing the input dataframe by naming the result of the query/drop operation something else. This would be straightforward if it weren't inside a loop. So I think I would do this (syntax may not be 100% correct):

```suggestion
    # one querystring with all cuts, joined by "or"
    querystring = " | ".join(f'(band == {bandname!r} & err > {cut})' for bandname, cut in nsigmaonmean.items())
    # name the result something else so that df_lc is unchanged
    clipped_df_lc = df_lc.drop(df_lc.query(querystring).index)
    # continue on using `clipped_df_lc`, then return it
```

Now when you (or the user) call the function, you are free to replace `df_lc` with the result if you want (`df_lc = sigmaclip_lightcurves(df_lc)`, and the main notebook does not need a new variable for every function), but you're not forced to change `df_lc` by the internal workings of the function. I think there are various performance and debugging reasons to prefer this, but it's also just nice for functions not to alter their input variables because that can be unexpected and confusing to a user who didn't want that.